### PR TITLE
IHub-175771 Update "ihub_sample" to showcase new approaches for creating an integration

### PR DESCRIPTION
**Pull request by glebowadim**

### CODE REVIEW

File: settings.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
 "UserName":"",
 "Password":"",
-"URL":"http://"
+"URL":"https://"
```
Comment (asmoiseenko) [inline settings.json:4]: Rename to OneVizionURL

File: settings.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
 "UserName":"",
```
Comment (asmoiseenko) [inline settings.json:2]: Rename to OneVizionUser

File: settings.json [hunk 3]
```diff
@@ -1,5 +1,5 @@
...
 "Password":"",
```
Comment (asmoiseenko) [inline settings.json:3]: rename to OneVizionSecurityToken

**Review by asmoiseenko — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -1,27 +1,29 @@
...
 
-Integration uses print(..., flush = True), so the stdout is synchronized with the Integration Logs. Otherwise, messages from stdout are added after the script is completed.
+Integration uses print(..., flush=True), so the stdout is synchronized with the Integration Logs. Otherwise, messages from stdout are added after the script is completed.
 
 ## Requirements
 - Python 3
-- Requests - library for python (http://docs.python-requests.org/en/master/)
+- Requests, library for python (http://docs.python-requests.org/en/master/)
+- onevizion library
+- jsonschema library
```
Comment (Anastasiyaaa) [inline README.md:23]: Maybe add links?

File: README.md [hunk 5]
```diff
@@ -1,27 +1,29 @@
...
 
-Sample Integration adds integration logs via API and writes message to stdout.
+Sample Integration adds integration logs via API and writes messages to stdout.
 
-Sample Integration consist of two python files, SimpleIntegration.py and OVIntegration.py.
-OVIntegration.py contains methods for working with API. SampleIntegration.py is python script for executing the integration. 
+Sample Integration consists of two python files, SimpleIntegration.py and OVIntegration.py.
+ovintegration.py contains a class for working with API. SampleIntegration.py is executable python script. 
 
 The integration adds logs of the different log levels (Info, Warning, Error, Debug) every 5 seconds.
```
Comment (Anastasiyaaa) [inline README.md:9]: I suggest adding a mention of the "Log Level" DropDown that iHub has (because logging is all that integration does)

**Review by Anastasiyaaa — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -1,27 +1,29 @@
...
+Integration uses print(..., flush=True), so the stdout is synchronized with the Integration Logs. Otherwise, messages from stdout are added after the script is completed.
 
 ## Requirements
 - Python 3
-- Requests - library for python (http://docs.python-requests.org/en/master/)
+- Requests, library for python (http://docs.python-requests.org/en/master/)
+- onevizion library
+- jsonschema library
 
 ## Usage
```
[the remark this answers is not in the document]
Comment (glebowadim) [reply README.md]: Removed

File: README.md [hunk 7]
```diff
@@ -1,27 +1,29 @@
 # ihub_sample
 Sample Project for Onevizion Integration Hub
 
-Sample Integration adds integration logs via API and writes message to stdout.
+Sample Integration adds integration logs via API and writes messages to stdout.
```
[the remark this answers is not in the document]
Comment (glebowadim) [reply README.md]: Added

File: SampleIntegration.py [hunk 8]
```diff
@@ -1,22 +1,41 @@
-import OVIntegration
-import json
 import time
```
[the remark this answers is not in the document]
Comment (glebowadim) [reply SampleIntegration.py]: Done

File: SampleIntegration.py [hunk 9]
```diff
@@ -1,22 +1,41 @@
...
+                               log_level_name=log_level)
 
-i = 0
-while i < 3:
+for i in range(3):
     time.sleep(5)
-    integrationOV = OVIntegration.OVIntegration(processId=processId, url=site, userName=user, password=password)
-    i = i + 1
-    print(i, flush = True)
+    ov_integration.add_log()
```
[the remark this answers is not in the document]
Comment (glebowadim) [reply SampleIntegration.py]: Moved to sample.py

File: SmallSampleIntegration.py [hunk 10]
```diff
@@ -1,7 +1,5 @@
 import time
 
-i = 0
-while i < 3:
+for i in range(3):
```
[the remark this answers is not in the document]
Comment (glebowadim) [reply SmallSampleIntegration.py]: I split all code into sample.py and start.py

File: settings.json (hunk 1 above)
Comment (glebowadim) [reply settings.json]: Ranamed

File: settings.json (hunk 2 above)
Comment (glebowadim) [reply settings.json]: Ranamed

File: settings.json (hunk 3 above)
Comment (glebowadim) [reply settings.json]: Renamed

Do you mean only renaming or should I use token authentication?

File: README.md (hunk 4 above)
Comment (glebowadim) [reply README.md]: Now this block is meaningless
Removed

File: README.md (hunk 5 above)
Comment (glebowadim) [reply README.md]: I don't think it's needed here. I think we can add this to the wiki

File: README.md [hunk 11]
```diff
@@ -1,40 +1,7 @@
...
-
-```json
-{
-"UserName":"amoiseenko",
-"Password":"******",
-"URL":"http://localhost:8080"
-}
-```
+Sample Integration consists of two python files, **start.py** and **sample.py**.
+**sample.py** contains a main business logic. **start.py** is an executable python script.
```
Comment (asmoiseenko) [inline README.md:5]: ... executable python script, which acts as a wrapper to read/validate configuration, install dependencies, and launche **sample.py**.

File: sample.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import time
+
+from onevizion import IntegrationLog, LogLevel
+
+
+class Integration:
+
+    def __init__(self, url: str, user_name: str, password: str, process_id: int, log_level_name: str):
+        url = url.replace("https://", "").replace("http://", "")
```
Comment (asmoiseenko) [inline sample.py:9]: I think it is bad idea to strip protocol part from the url parameter, propose to move this logic to the call of `IntegrationLog` constructor:
```
        self.integration_log_service = IntegrationLog(processId=process_id,
                                                      URL=url.replace("https://", "").replace("http://", ""),
                                                      userName=user_name,
                                                      password=password,
                                                      logLevelName=log_level_name)
```

File: sample.py [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+        self.integration_log_service = IntegrationLog(processId=process_id,
+                                                      URL=url,
+                                                      userName=user_name,
+                                                      password=password,
+                                                      logLevelName=log_level_name)
+
+    def _add_logs(self):
```
Comment (asmoiseenko) [inline sample.py:16]: I do not see any value of this routine, move single statement creating INFO level log entry to the start`s method loop

File: sample.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        self.integration_log_service.add(logLevel=LogLevel.INFO, message='Test1', description='TestD1')
+        self.integration_log_service.add(logLevel=LogLevel.DEBUG, message='Test2', description='TestD2')
+        self.integration_log_service.add(logLevel=LogLevel.WARNING, message='Test3', description='TestD3')
+        self.integration_log_service.add(logLevel=LogLevel.ERROR, message='Test4', description='TestD4')
+
+    def start(self):
+        for i in range(3):
+            time.sleep(5)
+            self._add_logs()
+            print(i, flush=True)
```
Comment (asmoiseenko) [inline sample.py:26]: why do you call `print` here?

File: sample.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
+import time
+
+from onevizion import IntegrationLog, LogLevel
+
+
+class Integration:
+
+    def __init__(self, url: str, user_name: str, password: str, process_id: int, log_level_name: str):
+        url = url.replace("https://", "").replace("http://", "")
+        self.integration_log_service = IntegrationLog(processId=process_id,
```
Comment (asmoiseenko) [inline sample.py:10]: Since log initialization is a common task for each integration, maybe it makes sense to move it to the `start.py`?

**Review by asmoiseenko — CHANGES_REQUESTED** [review]

File: settings.json (hunk 3 above)
Comment (asmoiseenko) [reply settings.json]: Use token - this is best preactise

File: README.md (hunk 11 above)
Comment (glebowadim) [reply README.md]: Done

File: sample.py (hunk 12 above)
Comment (glebowadim) [reply sample.py]: Done

File: sample.py (hunk 13 above)
Comment (glebowadim) [reply sample.py]: Done

File: sample.py (hunk 15 above)
Comment (glebowadim) [reply sample.py]: Done

File: settings.json (hunk 3 above)
Comment (glebowadim) [reply settings.json]: Done

File: sample.py (hunk 14 above)
Comment (glebowadim) [reply sample.py]: It was added to demonstrate "Read from stdout" feature. How to write log from stdout.
Not sure if this is needed here now.

**Comment by glebowadim** [comment]
@asmoiseenko 
Can you also mark the security hotspot as false positive?

File: start.py [hunk 16]
```diff
@@ -19,23 +19,22 @@
 except Exception as e:
     raise Exception("Incorrect value in the settings file\n {}".format(str(e))) from e
 
-user = settings_file_json["UserName"]
-password = settings_file_json["Password"]
-url = settings_file_json["URL"]
+access_key = settings_file_json["OneVizionAccessKey"]
```
Comment (asmoiseenko) [inline start.py:22]: Let's add some comments to highlight the purpose of each code block:
#read, validate, and parse integration settings file
#read standard integration environment properties
#initialize logging
#start the main integration code

Add comments to the `log_level`, ie:
```python
log_level = ihub_data['logLevel'] #retrieved from the integration`s settings in OneVizion
```

and `process_id`
```python
process_id = ihub_data['processId'] #the process_id of the current integration run in OneVizion
```

File: sample.py (hunk 14 above)
Comment (asmoiseenko) [reply sample.py]: Let's remove

**Review by asmoiseenko — CHANGES_REQUESTED** [review]

File: sample.py (hunk 14 above)
Comment (glebowadim) [reply sample.py]: Done

File: start.py (hunk 16 above)
Comment (glebowadim) [reply start.py:24]: Done

**Review by asmoiseenko — APPROVED** [review]

**Review by Anastasiyaaa — APPROVED** [review]

**Review by P4radise — APPROVED** [review]

**Pull request merged**